package.json

```diff
@@ -134,6 +134,7 @@
     "webpack-dev-server": "^3.11.2"
   },
   "peerDependencies": {
+    "graphql": "^15.0.0 || ^16.0.0",
```

Comment:
can graphql be an optional peer?  I'm curious if it's required for users who only define `rest` methods?

Reply:
That would be ideal, however it would require additional changes. Currently, the `graphql` package is included in the main entrypoint, so it's a required peer dependency of `msw`.

Reply:
I believe `graphql` should come in as an optional dependency. Otherwise, we'd be demanding our users who are not using GraphQL in their apps to install `graphql` alongside MSW. 

By making `graphql` an optional peer dependency (similar to how we list `typescript`), we are not making any breaking changes to our consumers, actually:

- Devs who are using GraphQL in their apps, already have GraphQL installed—no change here for them;
- Devs who are not using GraphQL are not affected at all.

Reply:
@kettanaito do you have an idea of how we might be able to make `graphql` optional? 

As far as I can tell, the only code being imported from `graphql` that is needed at runtime by `msw` is the `parse` function. It is used [here](https://github.com/mswjs/msw/blob/fc7f00c96e79ad0be83cbaf7c82ceff23f0c4355/src/utils/internal/parseGraphQLRequest.ts#L44).

My thought was to do something like this, to treat the package as optional:

```ts
// Since 'graphql' is an optional peer dependency,
// we'll attempt to import it dynamically
let parse: typeof import('graphql').parse
try {
  parse = require('graphql').parse
} catch (e) {}
```

What I'm unsure of is if the ESM output of `msw` will work with this approach.

The other option I considered is to move the code that depends on `graphql` to a different entrypoint (e.g. `msw/graphql`). This would be a breaking change though.

What do you think?

Reply:
If we have proper code splitting in place, if you're not using `graphql` from `msw`, your app shouldn't get the line of code where we do `import { parse } from 'graphql'`. This is something I'd kindly ask you to test out (steps described in some other comment on this pull request). 

And in terms of package.json, I believe you can take a look how we specify `"typescript"`, as it's an optional peer dependency we already have. 

> Side note: I think NPM also force-installs optional peer dependencies, which is sort of okay. As long as we don't break users' apps by forcing them to manually install a dependency they don't otherwise need.

Reply:
I'm not sure we can solve this with code splitting, without introducing a breaking change.

The problem is that the `graphql` import occurs within the main entrypoint of `msw`. When a consumer imports `msw`, and they don't have `graphql` in their `node_modules`, the `msw` entrypoint will fail to load because there will be a `MODULE_NOT_FOUND` error when trying to resolve `graphql`.

To code split, we would need to move the `graphql` import to a different entrypoint, which would be a breaking change.

I've opted to do the following in https://github.com/mswjs/msw/pull/1356/commits/7a8482ab8a1745533d9b845c0a0e6713eb79c2d1:
1. Mark `graphql` as an optional peer dependency
2. Dynamically load `graphql` within the `parseQuery` function using `require`. This function should only be called when the consumer is attempting to mock GraphQL with `msw` 
3. Update all static imports of `graphql` in the project to be `type imports`

I tested these changes by using my version of `msw` with the [with-msw](https://github.com/vercel/next.js/tree/canary/examples/with-msw) Next.js example. I was able to run the app, and mock rest requests, without needing to install `graphql`.

Reply:
I agree with your approach, I think it's great! Thanks for looking into this 👍